lib/Dialect/SPIRV/Transforms/DecorateSPIRVCompositeTypeLayoutPass.cpp

```diff
@@ -0,0 +1,316 @@
+//===- DecorateSPIRVCompositeTypeLayoutPass.cpp - Decorate composite type -===//
+//
+// Copyright 2019 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This file implements a pass to decoreate the composite types used by
+// composite objects in the StorageBuffer, PhysicalStorageBuffer, Uniform, and
+// PushConstant storage classes with layout information. See SPIR-V spec
+// "2.16.2. Validation Rules for Shader Capabilities" for more details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SPIRV/Passes.h"
+#include "mlir/Dialect/SPIRV/SPIRVDialect.h"
+#include "mlir/Dialect/SPIRV/SPIRVOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+/// According to the Vulkan spec "14.5.4. Offset and Stride Assignment":
+/// "There are different alignment requirements depending on the specific
+/// resources and on the features enabled on the device."
+///
+/// There are 3 types of alignment: scalar, base, extended.
+/// See the spec for details.
+///
+/// Note: Even if scalar alignment is supported, it is generally more
+/// performant to use the base alignment. So here the calculation is based on
+/// base alignment.
+///
+/// The memory layout must obey the following rules:
```

Comment:
These rules are not from the current Vulkan spec. Could you use the most recent version?

https://renderdoc.org/vkspec_chunked/chap14.html#interfaces-resources-layout

Reply:
@antiagainst oh, sorry, its a mix from Vulkan spec and SPIR-V spec.